From 9b2c77f01aab872d426d7b2cf1103423e700bd4b Mon Sep 17 00:00:00 2001
From: Divin Raj <divin.raj@arm.com>
Date: Thu, 17 Aug 2023 00:04:31 +0100
Subject: [PATCH] net: lib/sockets: Fix UDP remote address configuration
 issue

Configure UDP remote address to ensure packets are sent to the
correct network interface. Prevents potential issues when sending
data to secondary or subsequent interfaces using UDP.

Signed-off-by: Divin Raj <divin.raj@arm.com>
Upstream-Status: Pending
---
 subsys/net/lib/sockets/sockets.c | 26 ++++++++++++++++++++++++++
 1 file changed, 26 insertions(+)

diff --git a/subsys/net/lib/sockets/sockets.c b/subsys/net/lib/sockets/sockets.c
index 91bbe88d67..bf553316a5 100644
--- a/subsys/net/lib/sockets/sockets.c
+++ b/subsys/net/lib/sockets/sockets.c
@@ -1,6 +1,7 @@
 /*
  * Copyright (c) 2017 Linaro Limited
  * Copyright (c) 2021 Nordic Semiconductor
+ * Copyright (c) 2023, Arm Limited and Contributors. All rights reserved
  *
  * SPDX-License-Identifier: Apache-2.0
  */
@@ -525,6 +526,31 @@ int zsock_connect_ctx(struct net_context *ctx, const struct sockaddr *addr,
 			cb = zsock_connected_cb;
 		}
 
+		/* The binding must be performed after we have set the remote
+		 * address. Otherwise, address might not be properly set, which
+		 * could lead to problems when the UDP bind operation is subsequently
+		 * executed.
+		 *
+		 * For TCP, this scenario is managed within the net_context_connect()
+		 * function.
+		 */
+		if (IS_ENABLED(CONFIG_NET_UDP) &&
+				net_context_get_proto(ctx) == IPPROTO_UDP) {
+			if (IS_ENABLED(CONFIG_NET_IPV6) &&
+					net_context_get_family(ctx) == AF_INET6) {
+				struct sockaddr_in6 *addr6 = (struct sockaddr_in6 *)
+								&ctx->remote;
+				memcpy(&addr6->sin6_addr, &net_sin6(addr)->sin6_addr,
+					sizeof(struct in6_addr));
+			} else if (IS_ENABLED(CONFIG_NET_IPV4) &&
+					net_context_get_family(ctx) == AF_INET) {
+				struct sockaddr_in *addr4 = (struct sockaddr_in *)
+								&ctx->remote;
+				memcpy(&addr4->sin_addr, &net_sin(addr)->sin_addr,
+					sizeof(struct in_addr));
+			}
+		}
+
 		SET_ERRNO(net_context_recv(ctx, zsock_received_cb, K_NO_WAIT,
 					   ctx->user_data));
 		SET_ERRNO(net_context_connect(ctx, addr, addrlen, cb, timeout,
-- 
2.25.1

